community/modules/scripts/htcondor-install/files/autoscaler.py

```diff
@@ -185,6 +185,10 @@ def scale(self):
         # stewardship. A full list of Job ClassAd attributes can be found at
         # https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html
         schedd = htcondor.Schedd()
+        # encourage the job queue to start a new negotiation cycle
+        # there are internal unconfigurable rate limits so not guaranteed
+        schedd.reschedule()
+        time.sleep(3)
```

Comment:
All this really does is increase the odds that the filter on jobs allows very recently submitted jobs to pass. Thus the autoscaling is achieved with lower latency.

That said, the architecture of HTCondor is starting to allow the scheduler to do some of its own negotiating, especially when it rapidly processes a large queue of jobs with identical requests. In this case, the negotiator cycle is only part of the job-matching picture in the first place.

My top bullet for the next meeting with HTCondor team (early April) is how to best identify whether a job has participated in a negotiation cycle or direct scheduling by scheduler. I didn't find an obvious answer.

If we wanted to be very strict about this, we could repeatedly query the negotiator until `LastNegotiationCycleTime0` was recent. I *think* this might be nearly guaranteed of succeeding every 20 seconds, but it's a different form of brute force. In practice, at small scale, I didn't observe the sleep to be necessary to see the timestamp update within the same process.

Reply:
IIUC a natural negotiation cycle might happen every 20 seconds. You are forcing one to happen and then waiting, hoping that it has concluded. Is that correct? Is this really necessary for your scaling algorithm. It seems like the way you have it now where idle jobs and idle nodes cancel out that it wouldn't matter that there was not a recent negotiation.

To my original question, it seems like you are answering "no". I am just programed to always question a sleep in production code that is not tied to some state event. 

If you think we really need it then I would add a comment as to why 3. Otherwise it just looks like a magic number.

Reply:
I am removing the "magic number" (a fair provocation) in favor of a comment that clarifies that the encouragement of the negotiation cycle is nice, but not necessary.